\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage[]{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\PassOptionsToPackage{hyphens}{url} % url is loaded by hyperref
\usepackage[unicode=true]{hyperref}
\hypersetup{
            pdftitle={Assignment 4.3},
            pdfauthor={Mahdi},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother


\title{Assignment 4.3}
\author{Mahdi}
\date{11/3/2020}

\begin{document}
\maketitle

\subsubsection{We performed batch/species/colony effect correction for
the five ant species (A.ech, M.pha, S.inv, L.hum, and L.nig). Try to
include also the two queenless ant species in colony effect correction,
and report the similarity matrix (Heatmap+Hclustering that we used in
the classroom) and PCA result. (Code and Plots, and a briefly discussion
about what you see in the plot) (15
points)}\label{we-performed-batchspeciescolony-effect-correction-for-the-five-ant-species-a.ech-m.pha-s.inv-l.hum-and-l.nig.-try-to-include-also-the-two-queenless-ant-species-in-colony-effect-correction-and-report-the-similarity-matrix-heatmaphclustering-that-we-used-in-the-classroom-and-pca-result.-code-and-plots-and-a-briefly-discussion-about-what-you-see-in-the-plot-15-points}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Learn how to correct for batch effect.}
\CommentTok{# Learn how to identify differentially expressed genes across batch }
\KeywordTok{load}\NormalTok{(}\StringTok{'inputData.Rdata'}\NormalTok{)}
\KeywordTok{source}\NormalTok{(}\StringTok{"shared_functions.R"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: ggplot2
\end{verbatim}

\begin{verbatim}
## Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa
\end{verbatim}

\begin{verbatim}
## -- Attaching packages --------------------------------------- tidyverse 1.3.0 --
\end{verbatim}

\begin{verbatim}
## v tibble  3.0.4     v dplyr   1.0.2
## v tidyr   1.1.2     v stringr 1.4.0
## v readr   1.4.0     v forcats 0.5.0
## v purrr   0.3.4
\end{verbatim}

\begin{verbatim}
## -- Conflicts ------------------------------------------ tidyverse_conflicts() --
## x dplyr::filter() masks stats::filter()
## x dplyr::lag()    masks stats::lag()
\end{verbatim}

\begin{verbatim}
## Loading required package: ape
\end{verbatim}

\begin{verbatim}
## Loading required package: maps
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'maps'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:purrr':
## 
##     map
\end{verbatim}

\begin{verbatim}
## Loading required package: usethis
\end{verbatim}

\begin{verbatim}
## Loading required package: BiocGenerics
\end{verbatim}

\begin{verbatim}
## Loading required package: parallel
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'BiocGenerics'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:dplyr':
## 
##     combine, intersect, setdiff, union
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min
\end{verbatim}

\begin{verbatim}
## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.
\end{verbatim}

\begin{verbatim}
## Loading required package: S4Vectors
\end{verbatim}

\begin{verbatim}
## Loading required package: stats4
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'S4Vectors'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:dplyr':
## 
##     first, rename
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:tidyr':
## 
##     expand
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:base':
## 
##     expand.grid
\end{verbatim}

\begin{verbatim}
## Loading required package: IRanges
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'IRanges'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:dplyr':
## 
##     collapse, desc, slice
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:purrr':
## 
##     reduce
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:grDevices':
## 
##     windows
\end{verbatim}

\begin{verbatim}
## Loading required package: GenomicRanges
\end{verbatim}

\begin{verbatim}
## Loading required package: GenomeInfoDb
\end{verbatim}

\begin{verbatim}
## Loading required package: SummarizedExperiment
\end{verbatim}

\begin{verbatim}
## Loading required package: DelayedArray
\end{verbatim}

\begin{verbatim}
## Loading required package: matrixStats
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'matrixStats'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:Biobase':
## 
##     anyMissing, rowMedians
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:dplyr':
## 
##     count
\end{verbatim}

\begin{verbatim}
## Loading required package: BiocParallel
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'DelayedArray'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:matrixStats':
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:purrr':
## 
##     simplify
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     aperm, apply, rowsum
\end{verbatim}

\begin{verbatim}
## Loading required package: mgcv
\end{verbatim}

\begin{verbatim}
## Loading required package: nlme
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'nlme'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:IRanges':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:dplyr':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## This is mgcv 1.8-31. For overview type 'help("mgcv-package")'.
\end{verbatim}

\begin{verbatim}
## Loading required package: genefilter
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'genefilter'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:matrixStats':
## 
##     rowSds, rowVars
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:readr':
## 
##     spec
\end{verbatim}

\begin{verbatim}
## Registered S3 method overwritten by 'treeio':
##   method     from
##   root.phylo ape
\end{verbatim}

\begin{verbatim}
## ggtree v2.0.4  For help: https://yulab-smu.github.io/treedata-book/
## 
## If you use ggtree in published research, please cite the most appropriate paper(s):
## 
## [36m-[39m Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution 2018, 35(12):3041-3043. doi: 10.1093/molbev/msy194
## [36m-[39m Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'ggtree'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:nlme':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:IRanges':
## 
##     collapse
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:S4Vectors':
## 
##     expand
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:ape':
## 
##     rotate
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:tidyr':
## 
##     expand
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Section 2: Using Combat to normalize batch effect.}

\NormalTok{sampleTable}\OperatorTok{$}\NormalTok{caste[}\KeywordTok{which}\NormalTok{(sampleTable}\OperatorTok{$}\NormalTok{caste }\OperatorTok{==}\StringTok{ 'Minor_worker'}\NormalTok{)] =}\StringTok{ 'Worker'}  \CommentTok{# For simplity, we treat minor worker in A.echinator as worker caste}
\NormalTok{normal_ant =}\StringTok{ }\KeywordTok{which}\NormalTok{(sampleTable}\OperatorTok{$}\NormalTok{species }\OperatorTok{%in%}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"Aech"}\NormalTok{,}\StringTok{'Mpha'}\NormalTok{,}\StringTok{"Lhum"}\NormalTok{,}\StringTok{'Sinv'}\NormalTok{,}\StringTok{"Lnig"}\NormalTok{, }\StringTok{"Dqua"}\NormalTok{, }\StringTok{"Cbir"}\NormalTok{))}
\NormalTok{ortholog_exp.ant =}\StringTok{ }\NormalTok{ortholog_counts[,normal_ant]}
\NormalTok{sampleTable.ant =}\StringTok{ }\KeywordTok{droplevels}\NormalTok{(sampleTable[normal_ant,])}
\NormalTok{ortholog_exp.ant =}\StringTok{ }\NormalTok{ortholog_exp.ant[}\OperatorTok{!}\KeywordTok{apply}\NormalTok{(ortholog_exp.ant, }\DecValTok{1}\NormalTok{, anyNA),]  }\CommentTok{#Removed genes showing NA (e.g. without expression)}
\NormalTok{ortholog_exp.ant.norm =}\StringTok{ }\KeywordTok{log2}\NormalTok{(}\KeywordTok{normalize.quantiles}\NormalTok{(ortholog_exp.ant)}\OperatorTok{+}\DecValTok{1}\NormalTok{)}
\KeywordTok{colnames}\NormalTok{(ortholog_exp.ant.norm) =}\StringTok{ }\KeywordTok{colnames}\NormalTok{(ortholog_exp.ant)}
\KeywordTok{rownames}\NormalTok{(ortholog_exp.ant.norm) =}\StringTok{ }\KeywordTok{rownames}\NormalTok{(ortholog_exp.ant)}
\NormalTok{ortholog_exp.ant.norm =}\StringTok{ }\NormalTok{ortholog_exp.ant.norm[}\KeywordTok{apply}\NormalTok{(ortholog_exp.ant.norm, }\DecValTok{1}\NormalTok{, }
                                                    \DataTypeTok{FUN =} \ControlFlowTok{function}\NormalTok{(x) }\KeywordTok{return}\NormalTok{(}\KeywordTok{var}\NormalTok{(x, }\DataTypeTok{na.rm =}\NormalTok{ T) }\OperatorTok{>}\StringTok{ }\DecValTok{0}\NormalTok{)),] }

\NormalTok{batch =}\StringTok{ }\KeywordTok{droplevels}\NormalTok{(sampleTable.ant}\OperatorTok{$}\NormalTok{colony) }\CommentTok{# Normalization for species identity.}
\NormalTok{modcombat =}\StringTok{ }\KeywordTok{model.matrix}\NormalTok{(}\OperatorTok{~}\DecValTok{1}\NormalTok{, }\DataTypeTok{data =}\NormalTok{ sampleTable.ant)}
\NormalTok{combat.ortholog_exp.ant =}\StringTok{ }\KeywordTok{ComBat}\NormalTok{(}\DataTypeTok{dat=}\NormalTok{ortholog_exp.ant.norm, }\DataTypeTok{batch=}\NormalTok{batch, }\DataTypeTok{mod=}\NormalTok{modcombat,}
                             \DataTypeTok{mean.only =}\NormalTok{ F, }\DataTypeTok{par.prior=}\OtherTok{TRUE}\NormalTok{, }\DataTypeTok{prior.plots=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Found31batches
\end{verbatim}

\begin{verbatim}
## Adjusting for0covariate(s) or covariate level(s)
\end{verbatim}

\begin{verbatim}
## Standardizing Data across genes
\end{verbatim}

\begin{verbatim}
## Fitting L/S model and finding priors
\end{verbatim}

\begin{verbatim}
## Finding parametric adjustments
\end{verbatim}

\begin{verbatim}
## Adjusting the Data
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sampleDists.combat =}\StringTok{ }\KeywordTok{as.dist}\NormalTok{(}\DecValTok{1} \OperatorTok{-}\StringTok{ }\KeywordTok{cor}\NormalTok{(combat.ortholog_exp.ant,}\DataTypeTok{method =} \StringTok{'s'}\NormalTok{))}
\KeywordTok{pheatmap}\NormalTok{(sampleDists.combat,}\DataTypeTok{annotation_col =}\NormalTok{ sampleTable.ant[,}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{)], }
         \DataTypeTok{annotation_colors =}\NormalTok{ ann_colors,}
         \DataTypeTok{color =}\NormalTok{ colors) }
\end{Highlighting}
\end{Shaded}

\includegraphics{Assignment4.3-answers_files/figure-latex/unnamed-chunk-1-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{var.gene =}\StringTok{ }\KeywordTok{order}\NormalTok{(}\KeywordTok{apply}\NormalTok{(combat.ortholog_exp.ant,}\DecValTok{1}\NormalTok{,var),}\DataTypeTok{decreasing =}\NormalTok{ T)[}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\OperatorTok{:}\DecValTok{1000}\NormalTok{)]}
\NormalTok{ortholog_exp.combat.pca <-}\StringTok{ }\KeywordTok{PCA}\NormalTok{(}\KeywordTok{t}\NormalTok{(combat.ortholog_exp.ant[var.gene,]),}\DataTypeTok{ncp =} \DecValTok{4}\NormalTok{, }\DataTypeTok{graph =} \OtherTok{FALSE}\NormalTok{)}

\CommentTok{# Take a look at the amount of variations explained by each PC.}
\KeywordTok{fviz_eig}\NormalTok{(ortholog_exp.combat.pca, }\DataTypeTok{addlabels =} \OtherTok{TRUE}\NormalTok{,}\DataTypeTok{main =} \StringTok{'Explained variance for each PC'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{Assignment4.3-answers_files/figure-latex/unnamed-chunk-1-2.pdf}

\subsubsection{In the class, we tested the number of differentially
expressed genes between gyne and worker (castes) in one of the species
using DESeq2. Can you report the number of overlapping DEGs in two,
three, four, and all five typical ant species, i.e.~Aech Mpha Lhum Lnig
and Sinv, similar to the last slide of the class lecture? And how many
DEGs have you found if you use the model: Exp \textasciitilde{} Caste +
Species + Caste:Species ? Is this number different from the number of
overlapping DEGs? Why? (Number of DEGs for the two situations, and a
brief discussion about the difference) (15
points)}\label{in-the-class-we-tested-the-number-of-differentially-expressed-genes-between-gyne-and-worker-castes-in-one-of-the-species-using-deseq2.-can-you-report-the-number-of-overlapping-degs-in-two-three-four-and-all-five-typical-ant-species-i.e.aech-mpha-lhum-lnig-and-sinv-similar-to-the-last-slide-of-the-class-lecture-and-how-many-degs-have-you-found-if-you-use-the-model-exp-caste-species-castespecies-is-this-number-different-from-the-number-of-overlapping-degs-why-number-of-degs-for-the-two-situations-and-a-brief-discussion-about-the-difference-15-points}

\subsubsection{\texorpdfstring{Besides of using distance measuring
method, we can also use factor analysis approach to identify gene
modules. Try to extract the weight of genes from PC1 and PC2 of the five
ant species (normalized for species identify), and apply that to
transform the gene expression data on queenless ant species (Hint: Take
a look at
\url{http://www.iro.umontreal.ca/~pift6080/H09/documents/papers/pca_tutorial.pdf}
(Links to an external site.)). {[}Note: This is similar to train a gene
network from the training data set (the five ant species) and test it on
the target data set (the queenless ant).{]} (Code and Plot with the PC
data of both the five typical ants and the reconstructed PC data of the
two queenless ants, and a brief discussion about the result.) (3.4
points)}{Besides of using distance measuring method, we can also use factor analysis approach to identify gene modules. Try to extract the weight of genes from PC1 and PC2 of the five ant species (normalized for species identify), and apply that to transform the gene expression data on queenless ant species (Hint: Take a look at http://www.iro.umontreal.ca/\textasciitilde{}pift6080/H09/documents/papers/pca\_tutorial.pdf (Links to an external site.)). {[}Note: This is similar to train a gene network from the training data set (the five ant species) and test it on the target data set (the queenless ant).{]} (Code and Plot with the PC data of both the five typical ants and the reconstructed PC data of the two queenless ants, and a brief discussion about the result.) (3.4 points)}}\label{besides-of-using-distance-measuring-method-we-can-also-use-factor-analysis-approach-to-identify-gene-modules.-try-to-extract-the-weight-of-genes-from-pc1-and-pc2-of-the-five-ant-species-normalized-for-species-identify-and-apply-that-to-transform-the-gene-expression-data-on-queenless-ant-species-hint-take-a-look-at-httpwww.iro.umontreal.capift6080h09documentspaperspca_tutorial.pdf-links-to-an-external-site..-note-this-is-similar-to-train-a-gene-network-from-the-training-data-set-the-five-ant-species-and-test-it-on-the-target-data-set-the-queenless-ant.-code-and-plot-with-the-pc-data-of-both-the-five-typical-ants-and-the-reconstructed-pc-data-of-the-two-queenless-ants-and-a-brief-discussion-about-the-result.-3.4-points}

\end{document}
