---
title: "Assignment 2.2 Part 1"
author: "Mahdi"
date: "September 25, 2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Understanding the data

## Task 1
1. In the `sub_geno.tab` file, 0, 1 and 2 most likely represent the two homozygous and hgeteroguzous genotypes. -1 probably means missing data.
2. In the `sub_expr.tab` file, the rows are genes/transcripts and the columns are different samples and their gene expression values.
3. The `design.tab` file contains information about each column of the `sub_expr.tab` file. It says which population they belong to and other characteristics.

## Task 2
1. Calculate the number of missing genotypes for each SNP across all individuals.
```{r}
snps <- read.table("sub_geno.tab")
missing_count <- apply((snps == -1), 1, sum)
head(missing_count)
```

2. Calculate the minor allele frequency (MAF) for all SNPs across all individuals.
```{r}
get_geno_frequency <- function(geno, snp_mat){
  boolean_matrix <- snp_mat == geno
  count <- apply(boolean_matrix, 1, sum)
  freq <- count/dim(snps)[2]
  return (freq)
}

get_maf <- function(snp_mat){
  genotypes <- c(0,1,2)
  #get matrix of genotype frequencies
  geno_freq <- sapply(genotypes, get_geno_frequency, snp_mat)
  colnames(geno_freq) <- genotypes
  allele_freq <- geno_freq + geno_freq[,2]/2
  allele_freq <- allele_freq[,-2]
  maf <- apply(allele_freq, 1, min)
  return (maf)
}

maf <- get_maf(snps)
head(maf)
```

3. Filter our SNPs that have missing genotypes or a MAF<0.05 and use the filtered snps for the rest of the exercise.
```{r}
keep <- maf >= 0.05 & missing_count == 0
snps_filtered <- snps[keep,]
dim(snps_filtered)
```

4. Calculate the MAF for africans and non-africans separately. Is there a difference?
```{r}
design <- read.table("design.tab", header = T, sep = "\t")

filter_snp_population <- function(pop, snp_mat, design_mat, inv=F){
  if(inv){
    cols <- design_mat$Source.Name[design_mat$Characteristics.population. != pop]
  } else{
    cols <- design_mat$Source.Name[design_mat$Characteristics.population. == pop]
  }
  snp_mat <- (snp_mat[, cols])
  return(snp_mat)
}

african_snps <- filter_snp_population("YRI", snps_filtered, design)
non_african_snps <- filter_snp_population("YRI", snps_filtered, design, inv = T)

african_maf <- get_maf(african_snps)
non_african_maf <- get_maf(non_african_snps)
print("African")
print(head(african_maf))
print("Non-African")
print(head(non_african_maf))
```

Yes, there is a difference between the two groups.

## Task 3
1. Plot the distribution of expression levels across all samples for the ENSG00000172404.4 gene.
```{r}
gene_expr <- read.table("sub_expr.tab")
gene1 <- c("ENSG00000172404.4")
hist(as.matrix(gene_expr[gene1,]), main=gene1, xlab="Gene Expression Level")

```

2. Plot the expression levels of ENSG00000172404.4 against the genotypes of snp_22_41256802 and snp_22_45782142
```{r}
library(tidyverse)
snps1 <- c("snp_22_41256802", "snp_22_45782142")
gene1_col <- t(gene_expr[gene1,])
snps1_col <- t(snps_filtered[snps1,])
df <- cbind(snps1_col, gene1_col)
colnames(df) <- c(snps1, "expression")
df <- df %>% 
  as_tibble %>%
  gather(-expression, key="SNP", value ="genotype")
df$genotype <- as.factor(df$genotype)
df %>%
  ggplot(aes(x=genotype, y=expression)) + geom_boxplot() +
  facet_wrap(~SNP)

```

## Task 4
1. Do a linear regression of all sample genotypes on sample gene expression for snp_22_41256802 on ENSG00000172404.4
```{r}
lm_snp_a <- lm(gene1_col ~ snps1_col[,1])
summary(lm_snp_a)
```

2. Do a linear regression of all sample genotypes on sample gene expression for snp_22_45782142 on ENSG00000172404.4
```{r}
lm_snp_b <- lm(gene1_col ~ snps1_col[,2])
summary(lm_snp_b)
```

3. Make sense of the results (Understand what the values represent)
Gene expression levels can be somehwat explained by the first SNP but not the second. This is because the first SNP has an $R^2$ value of 0.5394 and very small p value for its coefficient meaning the model explains around half the variation in the data. In contrast the second SNP has an $R^2$ value of  0.002019 meaning the model explains almost no variation in the data.


## Question 4 Do a linear regression for snp_22_43336231 on ENSG00000100266.11
### Without covariates
```{r}
gene2 <- "ENSG00000100266.11"
snp2 <- "snp_22_43336231"

gene2_col <- t(gene_expr)[,gene2]
gene2_snp <- t(snps_filtered)[,snp2]
lm_no_cov <- lm(gene2_col ~ gene2_snp)
summary(lm_no_cov)
```

### Using the genotype PCs from pc_cvrt.tab as covariates
```{r}
pc <- read.table("data/pc_cvrt.tab")
lm_pc <- lm(gene2_col ~ gene2_snp + pc$PC1 + pc$PC2 + pc$PC3 + pc$PC4 + pc$PC5)
summary(lm_pc)
```

### Separately for african and non-africans without covariates. Hint: Use the information in the design.tab
```{r}
get_pop_gene <- function(genes, pop, gene_mat, design_mat, inv = F){
  genes_table <- filter_snp_population(pop, gene_mat, design_mat, inv)
  genes <- t(genes_table)[,genes]
  return(genes)
}

#make african model
gene2_africa <- get_pop_gene(gene2, "YRI", gene_expr, design)
snp2_africa <- t(african_snps)[,snp2]
lm_africa <- lm(gene2_africa ~ snp2_africa)
summary(lm_africa)

#make non african model
gene2_nonafrica <- get_pop_gene(gene2, "YRI", gene_expr, design, T)
snp2_nonafrica <- t(non_african_snps)[,snp2]
lm_nonafrica <- lm(gene2_nonafrica ~ snp2_nonafrica)
summary(lm_nonafrica)
```

### Make a dotplot of PC1 vs PC2 and color the dots by population
```{r}
library(tidyverse)
pc_df <- data.frame(PC1=pc$PC1, PC2=pc$PC2, Population=design$Characteristics.population.)
pc_df %>%
  #gather(-Population, key="PC", value="Value") %>%
  ggplot(aes(x=PC1, y=PC2, col=Population)) + 
  geom_point()
```

## Question 5
1. There is no difference since both models havea very low $R^2$ value. (probably wrong)
2. Both models are quite poor but the african model is much worse than the nona frican model since it has a lower $R^2$ value and a much higher p-value.
3 We are including the principal components of the gene expression data.

## Task 6 Do a linear regression on 1st snp on 1st gene, 2nd snp on 2nd gene etc.
1. Create a matrix containing the gene_id, snp_id, effect size, t.value and p.value
```{r}
gene_t <- as.data.frame(t(gene_expr))
snps_filtered_t <- as.data.frame(t(snps_filtered))

get_lm_values <- function(expr, snps){
  model <- summary(lm(expr ~ snps))
  return(model$coefficients[2,])
}

get_lm_matrix <- function(expr, snps, fun){
  lm_matrix <- mapply(fun, expr, snps)
  lm_matrix <- as.data.frame(t(lm_matrix))
  lm_matrix$gene_id <- rownames(lm_matrix)
  lm_matrix$snp_id <- colnames(snps)
  rownames(lm_matrix) <- NULL
  lm_matrix <- lm_matrix[,c(5,6,1,3,4)]
  colnames(lm_matrix) <- c("gene_id", "snp_id", "effect_size", "t.value", "p.value")
  return(lm_matrix)
}

lm_matrix <- get_lm_matrix(gene_t, snps_filtered_t, get_lm_values)
head(lm_matrix)
```

2. Do a multiple testing correction on the resulting p.values using fdr.
```{r}
lm_matrix$p.adj <- p.adjust(lm_matrix$p.value, method = "fdr")
lm_matrix %>%
  filter(p.adj < 0.05)
```

3. Do the same but now include the genotype PCs from pc_cvrt.tab as covariates.
```{r}
get_lm_values_covariate <- function(expr, snps){
  model <- summary(lm(expr ~ snps + pc$PC1 + pc$PC2 + pc$PC3 + pc$PC4 + pc$PC5))
  return(model$coefficients[2,])
}

lm_matrix_cov <- get_lm_matrix(gene_t, snps_filtered_t, get_lm_values_covariate)
lm_matrix_cov$p.adj <- p.adjust(lm_matrix_cov$p.value, method = "fdr")
lm_matrix_cov %>%
  filter(p.adj < 0.05)
```

4. Plot the most significant hit.
```{r}
sig_row <- lm_matrix_cov[lm_matrix_cov$p.adj == min(lm_matrix_cov$p.adj),]
df <- data.frame(Expression = gene_t[,sig_row$gene_id], 
                 Genotype = as.factor(snps_filtered_t[,sig_row$snp_id]))

df %>% 
  ggplot(aes(x=Genotype, y=Expression)) + geom_boxplot()
```

## Question 6
1. 32 tests were performed in a and 32 tests were performed in c.
2. Since multiple tests are being perfomred, it is possible to get a false positive by obtaining a significant p value by chance. The false positives are corrected using a multiple test correction. In eQTL analysis, a huge number of significance tests are performed so many flase posititves can occur, so it is essential to make sure to check for false positives.
3. No both models produced the same number of significant hits.


## Task 7
```{r}
library(MatrixEQTL)
snps <- SlicedData$new()
snps$CreateFromMatrix(as.matrix(snps_filtered)) #filt_geno is your filtered genotype matrix
genes <- SlicedData$new()
genes$CreateFromMatrix(as.matrix(gene_expr)) #expr is the unchanged expression matrix
snp_pos <- read.table("data/sample_geno.pos",sep="\t",header=T)
snp_pos <- snp_pos[snp_pos$snp %in% row.names(snps_filtered),]
gene_pos <- read.table("data/sample_expr.pos",sep="\t",header=T)
all(colnames(snps) == colnames(genes))
eQTL <- Matrix_eQTL_main(snps, genes, output_file_name=NULL,
output_file_name.cis=NULL,
pvOutputThreshold.cis=1, pvOutputThreshold=1,
snpspos=snp_pos, genepos=gene_pos,
cisDist = 0)
```

## Question 7
1. 1024 tests were performed.
2. MatrixeQTL found no significant hits while the analysis in Task 6 yielded 6 significant hits. This may be caused by


